pkg/hub/managedclusterset/controller.go

```diff
@@ -0,0 +1,265 @@
+	// no work to do if the cluster set is deleted
+	if !clusterSet.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	selectedClusters := []string{}
+	selectedCondition := clusterv1alpha1.StatusCondition{
+		Type: clusterv1alpha1.ManagedClusterSetConditionEmpty,
+	}
+
+	selectors := clusterSet.Spec.ClusterSelectors
+	if len(selectors) == 0 {
+		selectors = append(selectors, clusterv1alpha1.ClusterSelector{})
+	}
+
+	for _, selector := range selectors {
+		clusters, err := c.selectClusters(selector)
+		if err != nil {
+			return err
+		}
+
+		selectedClusters = append(selectedClusters, clusters...)
+	}
+
+	selectedClusters = unique(selectedClusters)
+	if count := len(selectedClusters); count == 0 {
+		selectedCondition.Status = metav1.ConditionTrue
+		selectedCondition.Reason = "NoClusterMatched"
+		selectedCondition.Message = "No ManagedCluster selected"
+	} else {
+		selectedCondition.Status = metav1.ConditionFalse
+		selectedCondition.Reason = "ClustersSelected"
+		selectedCondition.Message = fmt.Sprintf("%d ManagedClusters selected", count)
+	}
+
+	return c.updateClusterSetStatus(ctx, clusterSet, selectedCondition)
+}
+
+// selectClusters returns names of managed clusters which match the cluster selector
+func (c *managedClusterSetController) selectClusters(selector clusterv1alpha1.ClusterSelector) (clusterNames []string, err error) {
+	switch {
+	case len(selector.ClusterNames) > 0 && selector.LabelSelector != nil:
+		// return error if both ClusterNames and LabelSelector is specified for they are mutually exclusive
+		// This case should be handled by validating webhook
+		return nil, fmt.Errorf("both ClusterNames and LabelSelector is specified in ClusterSelector: %v", selector.LabelSelector)
+	case len(selector.ClusterNames) > 0:
+		// select clusters with cluster names
+		for _, clusterName := range selector.ClusterNames {
+			_, err = c.clusterLister.Get(clusterName)
+			switch {
+			case errors.IsNotFound(err):
+				continue
+			case err != nil:
+				return nil, fmt.Errorf("unable to fetch ManagedCluster %q: %w", clusterName, err)
+			default:
+				clusterNames = append(clusterNames, clusterName)
+			}
+		}
+		return clusterNames, nil
+	case selector.LabelSelector != nil:
+		// select clusters with label selector
+		labelSelector, err := convertLabels(selector.LabelSelector)
+		if err != nil {
+			// This case should be handled by validating webhook
+			return nil, fmt.Errorf("invalid label selector: %v, %w", selector.LabelSelector, err)
+		}
+		clusters, err := c.clusterLister.List(labelSelector)
+		if err != nil {
+			return nil, fmt.Errorf("unable to list ManagedClusters with label selector: %v, %w", selector.LabelSelector, err)
+		}
+		for _, cluster := range clusters {
+			clusterNames = append(clusterNames, cluster.Name)
+		}
+		return clusterNames, nil
+	default:
+		// select all clusters if neither ClusterNames nor LabelSelector is specified
+		clusters, err := c.clusterLister.List(labels.Everything())
```

Comment:
The admission plugin needs to correctly handle this right?

Reply:
Hrm, I am not sure if I would prefer an empty spec to mean "everything" or "empty". I would probably prefer "empty", which I also think is a valid construct. I feel like if you want to make a set that matches everything, you should have to say so explicitly, since I would view this as a relatively infrequent case.

Reply:
Updated to select no cluster if neither ClusterNames nor LabelSelector is specified. And I'll create an another PR to update comments on CRD.